example/cpp2/include/ICU4XDataProvider.d.hpp

```diff
@@ -21,10 +21,10 @@ class ICU4XDataProvider {
 
   inline static diplomat::result<std::monostate, std::monostate> returns_result();
 
-  inline const capi::ICU4XDataProvider* AsFFI() const;
-  inline capi::ICU4XDataProvider* AsFFI();
-  inline static const ICU4XDataProvider* FromFFI(const capi::ICU4XDataProvider* ptr);
-  inline static ICU4XDataProvider* FromFFI(capi::ICU4XDataProvider* ptr);
+  inline const ::capi::ICU4XDataProvider* AsFFI() const;
```

Comment:
Issue: I don't like using `::` on headers meant for inclusion; it prevents someone from including this inside their own namespace. For example, someone can do

```cpp
namespace my_icu4x {
    #include "icu4x/ICU4XDataProvider.h"
}
```

and it should "just work".

Reply:
No, that's extremely poor practice and leads to all the C++ `#include`s ending up in the wrong namespace. It doesn't work.

You should basically never be invoking `#include` anywhere but the top level of the file, unless it's one of those polymorphic headers explicitly designed to be invoked inline.

Reply:
I don't think we _need_ the `::` though?

I've definitely done this pattern before. It's a useful way for an intermediate library to include a private dependency.

Reply:
The idea is that you can make `my_icu4x.h` with

```cpp
namespace my_icu4x {
    #include "icu4x/ICU4XDataProvider.h"
}
```

and now you can `#include "my_icu4x.h"` without polluting your namespace.

Reply:
It doesn't work, and when it does it's extremely brittle and does not work for long.

Every non-ICU4X `#include` will get polluted this way. This will mean that things from `#include<iostream>` will end up in `my_icu4x` too, _breaking the rest of the code_.

The only way it can work is if we do not use any stdlib headers. But we do, and we're not going to stop doing that.

See https://softwareengineering.stackexchange.com/questions/298576/c-extensible-namespaces-how-to-force-declarations-back-into-global-namespace , https://softwareengineering.stackexchange.com/questions/335240/is-including-headers-inside-a-namespace-always-a-bad-idea

Reply:
ok

Reply:
There's still an issue I think: the namespace isn't `::capi::ICU4XDataProvider`, it is `::icu4x::capi::ICU4XDataProvider`, right? Is that part not implemented yet?

Reply:
Also,

> I don't think we _need_ the `::` though?

Separately from the use case I raised which you've convinced me is a bad practice we shouldn't care to support, resorting to `::` seems like a lazy solution to a problem that should be better solved by the code keeping track of what namespace the types are going in.

Reply:
Not in our Diplomat example it isn't. In the Diplomat tests only one or two things are namespaced.

Reply:
I can update the example, we haven't been keeping it polished

Reply:
> Not in our Diplomat example it isn't. In the Diplomat tests only one or two things are namespaced.

Can you point me to the files that demonstrate the non-default namespaces?

Reply:
I already updated the example code to do this.

But the feature_tests `Attr` tests are mostly namespaced under `ns`